File: leet-code/787. Cheapest Flights Within K Stops/CheapestFlightsWithinKStops787_2.java
package com.example;

import java.util.*;

public class CheapestFlightsWithinKStops787_2 {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        Map<Integer, List<int[]>> map = new HashMap<>();
        for (int[] flight : flights) {
            int start = flight[0];
            int end = flight[1];
            int cost = flight[2];
            map.putIfAbsent(start, new ArrayList<>());
            map.get(start).add(new int[]{end, cost});
        }

        int[] stops = new int[n];
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.add(new Node(src, 0, 0));

        while (!pq.isEmpty()) {
            Node departure = pq.poll();

            if (departure.from == dst) {
                return departure.price;
            }

            stops[departure.from] = departure.stop;

            if (departure.stop > k) {
                continue;
            }

            if (!map.containsKey(departure.from)) {
                continue;
            }

            for (int[] node : map.get(departure.from)) {
                int arrival = node[0];
                int price = node[1];
                // 비용은 고려하지 않는다
                if (stops[arrival] == 0 || (stops[arrival] > 0 && stops[arrival] > departure.stop + 1)) {
                    // stops[arrival] = departure.stop + 1;
                    // stops 를 여기서 갱신하면 안 된다
                    // 쉽게 생각하면 보통 다익스트라를 구현할 때 여기서 방문 처리를 하지 않는다
                    // 우선순위 큐에서 꺼내기도 전에 stops 가 갱신되어 버린다
                    // 반례 -> (leetcode 예시 2) 3, [[0, 1, 100], [1, 2, 100], [0, 2, 500]], 0, 2, 1
                    pq.add(new Node(arrival, departure.price + price, departure.stop + 1));
                }
            }
        }

        return -1;
    }

    private static class Node implements Comparable<Node> {
        int from;
        int price;
        int stop;

        @Override
        public int compareTo(Node o) {
            return this.price - o.price;
        }

        Node(int from, int price, int stop) {
            this.from = from;
            this.price = price;
            this.stop = stop;
        }
    }
}

File: leet-code/787. Cheapest Flights Within K Stops/CheapestFlightsWithinKStops787_2.kt
package com.example

import java.util.PriorityQueue

class CheapestFlightsWithinKStops787_2 {
    // 일반적으로 풀이했던 다익스트라와 달리
    // 방문한 정점을 다시 방문할 수 있도록 하는 조건이 필요하다
    fun findCheapestPrice(n: Int, flights: Array<IntArray>, src: Int, dst: Int, k: Int): Int {
        // 인접 리스트로 정점별 방문 가능한 정점 및 비용 정보
        val map = mutableMapOf<Int, MutableList<IntArray>>()
        for (flight in flights) {
            val start = flight[0]
            val end = flight[1]
            val price = flight[2]
            map.putIfAbsent(start, mutableListOf())
            map[start]!!.add(intArrayOf(end, price))
        }

        // 정점별 경유 횟수
        val stops = IntArray(n)
        // 출발지, 비용, 경유 횟수
        val pq = PriorityQueue<IntArray>{ o1, o2 -> o1[1] - o2[1]}
        pq.add(intArrayOf(src, 0, 0))

        while (!pq.isEmpty()) {
            val departure = pq.poll()

            if (departure[0] == dst) {
                return departure[1]
            }

            if (departure[2] > k) {
                continue
            }

            // 경유 횟수 갱신
            // (일종의 방문 처리)
            stops[departure[0]] = departure[2]

            if (!map.containsKey(departure[0])) {
                continue
            }

            for (arrival in map[departure[0]]!!) {
                // stops 를 기준으로 보면
                // 0 -> 1 -> 2 로 가는 경로를 거치면
                // 정점 2의 경유 횟수는 2가 된다
                // 그리고나서
                // 0 -> 2 로 가는 경로가 있으면
                // 정점 0의 경유 횟수는 0이라
                // 정점 0의 경유 횟수에 1을 더한 것보다 정점 2의 경유 횟수가 더 크다
                if (stops[arrival[0]] == 0 || stops[arrival[0]] > departure[2] + 1) {
                    pq.add(intArrayOf(arrival[0], departure[1] + arrival[1], departure[2] + 1))
                }
            }
        }

        return -1
    }
}